component/stream/read_write.go

```diff
@@ -0,0 +1,197 @@
+package stream
+
+import (
+	"blobfuse2/common"
+	"blobfuse2/common/log"
+	"blobfuse2/internal"
+	"blobfuse2/internal/handlemap"
+	"errors"
+	"io"
+	"sync/atomic"
+
+	"github.com/pbnjay/memory"
+)
+
+type ReadWriteCache struct {
+	*Stream
+	StreamConnection
+	bufferSizePerHandle uint64 // maximum number of blocks allowed to be stored for a file
+	handleLimit         int32
+	cachedHandles       int32
+	streamOnly          bool
+}
+
+func (rw *ReadWriteCache) Configure(conf StreamOptions) error {
+	if conf.BufferSizePerFile <= 0 || conf.HandleLimit <= 0 {
+		rw.streamOnly = true
+	}
+	rw.bufferSizePerHandle = conf.BufferSizePerFile
+	rw.handleLimit = int32(conf.HandleLimit)
+	rw.cachedHandles = 0
+	return nil
+}
+
+func (rw *ReadWriteCache) OpenFile(options internal.OpenFileOptions) (*handlemap.Handle, error) {
+	log.Trace("Stream::OpenFile : name=%s, flags=%d, mode=%s", options.Name, options.Flags, options.Mode)
+	handle, err := rw.NextComponent().OpenFile(options)
+	if err != nil {
+		log.Err("Stream::OpenFile : error %s [%s]", options.Name, err.Error())
+		return handle, err
+	}
+	if handle == nil {
+		handle = handlemap.NewHandle(options.Name)
+	}
+	if !rw.streamOnly {
+		if rw.cachedHandles >= rw.handleLimit {
+			log.Trace("Stream::OpenFile : file handle limit exceeded - switch handle to stream only mode %s [%s]", options.Name, handle.ID)
+			handle.CacheObj.StreamOnly = true
+			return handle, nil
+		}
+		atomic.AddInt32(&rw.cachedHandles, 1)
+		handlemap.CreateCacheObject(int64(rw.bufferSizePerHandle), handle)
+		opts := internal.GetFileBlockOffsetsOptions{
+			Name: handle.Path,
+		}
+		offsets, _ := rw.GetFileBlockOffsets(opts)
+		offsets.Cached = true
+		handle.CacheObj.BlockOffsetList = offsets
+	}
+	return handle, err
+}
+
+func (rw *ReadWriteCache) ReadInBuffer(options internal.ReadInBufferOptions) (int, error) {
+	if rw.streamOnly || options.Handle.CacheObj.StreamOnly {
+		data, err := rw.NextComponent().ReadInBuffer(options)
+		if err != nil && err != io.EOF {
+			log.Err("Stream::ReadInBuffer : error failed to download requested data for %s: [%s]", options.Handle.Path, err.Error())
+		}
+		return data, err
+	}
+	return rw.readWriteBlocks(options.Handle, options.Offset, options.Data, false)
+}
+
+func (rw *ReadWriteCache) WriteFile(options internal.WriteFileOptions) (int, error) {
+	log.Trace("Stream::WriteFile : name=%s, handle=%d", options.Handle.Path, options.Handle.ID)
+	if rw.streamOnly || options.Handle.CacheObj.StreamOnly {
+		data, err := rw.NextComponent().WriteFile(options)
+		if err != nil && err != io.EOF {
+			log.Err("Stream::WriteFile : error failed to write data for %s: [%s]", options.Handle.Path, err.Error())
+		}
+		return data, err
+	}
+	data, err := rw.readWriteBlocks(options.Handle, options.Offset, options.Data, true)
+	return data, err
+}
+
+func (rw *ReadWriteCache) CloseFile(options internal.CloseFileOptions) error {
```

Comment:
Just like our main component we shall have both FlushFile and CloseFile seperatly. In flush just upload and in close call flush and then close the handle, may be delete all cached buffers and all. However in flush we keep our buffers and just upload the current changes.